### ISSUE TITLE
sh subprocess replacement doesn't work on Windows

### ISSUE THREAD

**Issue by skral** [nerdvegas/rez#308]
Tried to release a package on Windows with git. Was able to workaround the issue by using pbs but more of a workaround than a real solution.
sh isn't available for Windows. pbs (predecessor to sh) is but is sh or pbs really necessary? subprocess should do the job just fine.

sh is being used in [rezplugins/release_hook/command.py](https://github.com/nerdvegas/rez/blob/master/src/rezplugins/release_hook/command.py#L12) and [rezplugins/release_vcs/git.py](https://github.com/nerdvegas/rez/blob/master/src/rezplugins/release_vcs/git.py#L7)

I would like to replace sh but wanted to check in what you think about the issue?

**Comment by mstreatfield** [comment]
We have been able to release from Windows using Git in the past and not had
this problem, so perhaps it's a recent regression? Although I've not been
doing Rez dev for a while so might be wrong.

One thing we found really useful where I used to work was to setup
Jenkins/Hudson/etc to build rez and run the test suite on Windows, Linux
and OSX whenever we pulled changes from the external repo (we actually did
it on every commit to every branch). It helped catch these changes before
they got to production as I think Allan is only able to test on Linux.

Sorry, not a solution I know, but thought it might be useful.

Mark

**Comment by nerdvegas** [comment]
Hey Sebastian,

Thanks for the head's up, looking at the sh code just now I see that it
explicitly doesn't support Windows.

I'm open to replacing it but I'd do a bit of investigation first to see if
anything similar (but cross platform) could take it's place, as it's pretty
convenient. If not, it's probably worth adding an sh-like interface in rez
ourselves, it doesn't strike me as difficult to do.

A

**Comment by nerdvegas** [comment]
Hey Mark,

Do you recall at all how you guys did this? There's an explicit ImportError
raised in sh.py on Windows.

Cheers
A

**Comment by mstreatfield** [comment]
@nerdvegas no, sorry.  I was definitely able to do a release with a git tag from Windows, and I do remember I had to work around some `sh` module issues, but I don't recall what exactly (and probably would have pushed them to a pull request at some point).

I don't have access to the original source code - @fnaum might be able to check.

**Comment by instinct-vfx** [comment]
I'd like to play with this, as it would be a hard requirement for us. I was wondering what everyone's preferred ways to tackle this would be. Is there a wish for something like sh/pbs or similar to have similar implementations in across the different VCS plugins? Obvious possibilities i currently see:

- Write something similar that works across platforms (daunting for me, as i do neither OSX nor Linux)
- Use any of the existing git modules (GitPython, pygit, Dulwich, Gittle, )
- Replace sh with subprocess calls (or write a thin subprocess based wrapper)
- ?

As said i'd be willing to spend some time to experiment a little as it would be a requirement for us and i'd like to see that as part of the official distribution should we go for rez. I would just prefer to spend the time on something that works for others.

Cheers,
Thorsten

**Comment by nerdvegas** [comment]
We started with GitPython. I ditched it because it's far, far easier to
find help on any git issues you run into, using cli interface. GitPython
just made it more difficult to fix anything.

I am in favor of writing a thin, sh-like wrapper based on sh. That way it
can be reused in other cases where we need to subproc (build system for
eg), rather than just giving us a solution for git alone.

A

**Comment by skral** [comment]
I think an sh-like wrapper sounds like a good idea, especially for the output parsing and exit codes. Why not try [pbs](http://pypi.python.org/pypi/pbs) (the sh predecessor)? It looks like it's supporting all platforms and on a quick look it has all the features rez used from sh.
pbs is under the MIT license, so we are able to fork it and improve on it if we really have to.

**Comment by mvtk** [comment]
pbs seems like a sensible choice to me. Will see if i find the time soon to cook something up.

**Comment by instinct-vfx** [comment]
Sorry. Wrong account again. *sigh*

**Comment by nerdvegas** [comment]
If pbs is sh-like without the linux constraint then that sounds like a good
candidate to me.

A

**Comment by instinct-vfx** [comment]
So it seems that sh is not used anymore anyways? All i did was remove the import in the git plugin and i could realease just fine. Looking into the code this seems to be based on subprocess already.

See https://github.com/nerdvegas/rez/blob/master/src/rez/release_vcs.py#L201

subprocess is used to do the actual call. The import in the gitplugin seems unused.

**Comment by nerdvegas** [comment]
Well there you go. There must have been a reason sh was removed but I don't
recall what it was.

A


On Thu, Apr 14, 2016 at 5:18 AM, Thorsten Kaufmann <notifications@github.com

**Comment by fnaum** [comment]
Hi, sorry for jumping late.

I just check the changes we have here and the one that Mark submitted on the PRs

https://github.com/nerdvegas/rez/pull/159
https://github.com/nerdvegas/rez/pull/176

I can confirm that we can release on windows and the git part work fine tags are applied etc.

**Comment by skral** [comment]
@instinct-vfx Bringing this topic back up after you were talking about it on the mailing-list. Did you have a chance to rework the sh part?

**Comment by nerdvegas** [comment]
Hey Sebastian,

Not a chance yet no, my time is very limited nowdays. I get 2-3 hours a
week dev time outside of work, if I'm lucky (although sometimes there are
work requested features which get done on work time).

A

**Comment by instinct-vfx** [comment]
@skral So there is two parts. to this. Since 2.0.0 (or 2.1.0? not sure) there was an additional piece of code added that uses sh at the end of the git plugin in the export method. For the rest of the module it is enough to drop the import. For the export method i can not use the self.git() implementation the rest of the class uses as the method is marked as a classmethod and hence can't access self.

@nerdvegas : Does this have to be a classmethod? Then i'd have to duplicate a bit of code from the class in the method to use the same approach throughout the module.

Or did you mean something different Sebastian?

Thorsten
